Array destructuring

romdotdog opened the pull request:
- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

This is a ~~work in progress~~ ~~almost finished~~ finished PR to hopefully implement array destructuring

romdotdog commented on src/compiler.ts:
```diff
@@ -2999,195 +3000,325 @@ export class Compiler extends DiagnosticEmitter {
+                }
+
+                let callExpr = this.compileCallDirect(indexedGet, [
+                  Node.createIntegerLiteralExpression(i64_new(j), nameNode.range)
```
Is there a better way to do this?

dcodeIO replied:
Perhaps `makeCallDirect` to circumvent the AST and provide a Binaryen IR node directly :)

romdotdog commented:
Should I add rest parameters? If so, should I rely on there being a `slice` method like in 
https://github.com/AssemblyScript/assemblyscript/issues/1473#issuecomment-699172441

trusktr commented:
:partying_face:

romdotdog commented:
Parser test `parser/continue-on-error.ts` was failing, now fixed

romdotdog commented:
To reduce the size of this PR, I will not be including rest parameters or array destructuring in function parameters
This PR is now generally ready for review and merge, I will meanwhile make sure that tests pass

MaxGraey commented on src/extra/ast.ts:
```diff
@@ -613,13 +638,13 @@ export class ASTBuilder {
     var declaration = node.declaration;
     if (!declaration.arrowKind) {
-      if (declaration.name.text.length) {
+      if (declaration.ident.text.length) {
```
Hmm, I'm not sure `ident` is good name. It could be misreading as `indent`. Why not keep the old `name`?

romdotdog replied:
What should the old `name` (that is still in use) become?

MaxGraey replied:
I mean revert it to `declaration.name`

romdotdog replied:
> 
> 
> I mean revert it to `declaration.name`

Can't, `ident` is a getter function, `name` is still a property of DeclarationStatement

romdotdog replied:
Solved by discussion below

MaxGraey replied:
I see. How about `id` of full name `identifier` instead `ident`?

MaxGraey commented:
Does swapping support?
```ts
let a = 1;
let b = 3;
[a, b] = [b, a];
assert(a == 3);
assert(b == 1);
```

romdotdog commented:
Yes, it does.

dcodeIO commented on src/ast.ts:
```diff
@@ -2045,7 +2073,7 @@ export class FunctionDeclaration extends DeclarationStatement {
   clone(): FunctionDeclaration {
     return new FunctionDeclaration(
-      this.name,
+      this.name as IdentifierExpression,
```
This can fail silently when compiled to JS. Adding an `assert` above to make sure that the `kind` matches would unify the behavior.

Can you explain why `name` needs to be `Expression`?

romdotdog replied:
It needs to be able to hold BindingPatternExpression

romdotdog replied:
It's really a `BindingPatternExpression | IdentifierExpression`

dcodeIO replied:
Hmm, seems odd that this propagates up to `FunctionDeclaration`. Is there maybe a way to separate this?

romdotdog replied:
no, because `DeclarationStatement` always holds the property `name`
I recall that @jtenner also thinks this is the only way in a voice chat we had

dcodeIO replied:
What if the `name` property would become a getter that can be overridden with a more specific type where necessary, say with a private `_name` that is `Expression`? Would that work?

romdotdog replied:
That's a good idea, I'll implement that after I have breakfast

romdotdog commented on src/ast.ts:
```diff
@@ -1701,6 +1724,17 @@ export abstract class DeclarationStatement extends Statement {
+  /** Attempts to return the name as an identifier. */
+  get ident(): IdentifierExpression {
+    assert(this._name.kind == NodeKind.IDENTIFIER || this._name.kind == NodeKind.CONSTRUCTOR);
+    return this._name as IdentifierExpression;
```
any better name for this function?

dcodeIO replied:
Perhaps `identifier`, we don't abbreviate so much elsewhere if I'm not mistaken.

romdotdog commented on src/program.ts:
```diff
@@ -2897,7 +2927,7 @@ export abstract class DeclaredElement extends Element {
   get identifierNode(): IdentifierExpression {
-    return this.declaration.name;
+    return this.declaration.ident;
   }
```
This seems wrong

romdotdog commented on src/program.ts:
```diff
@@ -2571,6 +2569,27 @@ export class Program extends DiagnosticEmitter {
+  private initializeVariable(
+    /** Name of the variable to initialize. */
+    name: string,
+    /** Parent element, usually a file or namespace. */
```
I should probably inline this function again

MaxGraey commented:
Add please tests for empty destructors:
```ts
const arr = [1, 2, 3];
const [] = arr;
```
and
```ts
const arr = [1, 2, 3];
const [,] = arr;
```

romdotdog commented:
good catch, thank you

MaxGraey commented:
Fo next PR will be great also add default assignment as well:
```ts
const [a = 1] = [];
assert(a == 1);
```

romdotdog commented:
@MaxGraey 

> For now without default value this `const [a] = [] as i32[];` should throw OOB trap
It does. Is there a way integrate this into the test? (test for traps?)

Also, in general, does this PR look good to you?

MaxGraey commented:
It look good to me but better if @dcodeIO review this

MaxGraey commented:
@romdotdog btw in this PR `package-lock.json` was changed. Could you remove this unnecessary changes?

romdotdog commented:
That's weird, it seems the features test failed because of the nvm GitHub action.
cc: @dcodeIO

MaxGraey commented:
> That's weird, it seems the features test failed because of the nvm GitHub action.

Yeah. That's sometimes happening with nightly builds. Don't care about this. Just need some time

MaxGraey commented:
> That's weird, it seems the features test failed because of the nvm GitHub action.
> cc: @dcodeIO

please sync with main

dcodeIO commented on src/ast.ts:
```diff
@@ -55,6 +55,7 @@ export enum NodeKind {
 
   // expressions
   IDENTIFIER,
+  PATTERN,
```
What about `BINDINGPATTERN`? :)

romdotdog replied:
Done @dcodeIO

romdotdog commented:
> 
> 
> > That's weird, it seems the features test failed because of the nvm GitHub action.
> > cc: @dcodeIO
> 
> please sync with main

Are you talking to me? Is there any trouble with merging this? GitHub is telling me that there are no conflicts :)

MaxGraey commented:
Yeah, no need, that's why I hid that message. By the way, have you seen the other dcodeIO suggestions?
https://github.com/AssemblyScript/assemblyscript/pull/2008#discussion_r685821303
https://github.com/AssemblyScript/assemblyscript/pull/2008#discussion_r685818957

romdotdog commented:
Thank you! I completely missed those. I'll implement those tomorrow

MaxGraey commented:
Could you also add destruct tests for typed and static arrays?
```ts
let arr = new Float32Array(3);
arr[0] = 1;
arr[1] = 2;
arr[2] = 3
let [a, b, c] = arr;
assert(a == 1);
assert(b == 2);
assert(c == 3);
```

MaxGraey commented on tests/parser/array-destructuring.ts:
```diff
@@ -0,0 +1,6 @@
+function test1(): void {
+  const [, a, b] = c;
+  const {d, e} = f;
+  const [] = g;
+  const [,] = h;
```
Add please more tests here. Like:
```ts
const [,,  ,
,] = k;
let [
] = n;
var [,a,] = j
```

romdotdog commented:
BTW, for the next PR I think I'm going to replace binding patterns with array literals for more flexible parsing and for normal assignments and declarations to have more similar code

MaxGraey commented:
@romdotdog It seems some tests have failed

MaxGraey commented on tests/parser/array-destructuring.ts:
```diff
@@ -3,4 +3,21 @@ function test1(): void {
+  ] = n;
+
+  var [,_,] = j ;
+  var [,,  ,a,, f ,
+    , g,  g,        ] = 1;
+  var [n, a, a, a,v,
+    ,h,h,h,h,h,,r
+    ,,r,,h
+    ,y, ] = 2;
+  let {aaa , a} = 1;
+  var {f,,,,,,,,,,,,} = 3;
+  const {fff} = 3;
```
Btw all that should an error like:
```
TS2488: Type 'number' must have a iterator.
```

romdotdog replied:
Yes, but it's a parser test, not a compiler test

romdotdog replied:
If this is a request I'll add a test for that but it's unclear what you're asking :)

dcodeIO commented:
I like the idea of reusing array literal expressions here, would be cool if it could be done right away :)

romdotdog commented:
> 
> 
> I like the idea of reusing array literal expressions here, would be cool if it could be done right away :)

It's an easily fixable issue; I don't think this PR is worth delaying more for something that can be included in the next PR (:

romdotdog commented:
BTW, I couldn't reproduce the failed compiler test on my machine.
Not to mention that it failed without me making any changes to the compiler test or the compiler itself.. Maybe someone could point me in the right direction?

romdotdog commented:
Had some extra time, pulled from upstream and squashed the resulting bugs.
I also removed the custom BindingPatternExpression and replaced it with array and object literals, as requested.

I hope this can be merged soon!

romdotdog commented:
Closing after the talk on [Discord](https://discord.com/channels/721472913886281818/874686570148536420/914344796431216670) about the `.identifierNode` 'hack' used throughout this PR, view by joining the [Discord Server](https://discord.gg/assemblyscript)
